pandas/core/arrays/integer.py

```diff
@@ -333,7 +337,19 @@ def _from_sequence_of_strings(
     def _coerce_to_array(self, value) -> tuple[np.ndarray, np.ndarray]:
         return coerce_to_array(value, dtype=self.dtype)
 
-    def astype(self, dtype, copy: bool = True) -> ArrayLike:
+    @overload
+    def astype(self, dtype: npt.DTypeLike, copy: bool = ...) -> np.ndarray:
```

Comment:
I'm not sure whether the following is possible:

```py
T = TypeVar("T", npt.DTypeLike)
def astype(self, dtype: T, copy: bool = ...) -> npt.NDArray[T]
```

Might be interesting to compare that to numpy's ndarray.astype signature.

Reply:
> I'm not sure whether the following is possible:

Doesn't work:
```python
>>> import numpy.typing as npt
>>> from typing import TypeVar
>>> T = TypeVar("T", npt.DTypeLike)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Anaconda3\envs\pandas-dev-types\lib\typing.py", line 596, in __init__
    raise TypeError("A single constraint is not allowed")
TypeError: A single constraint is not allowed
```

Reply:
That's how numpy does it. It is a  bit more complex:
https://github.com/numpy/numpy/blob/1fe57b2427e20f466c254ed6a98a014d9a9e574e/numpy/__init__.pyi#L1920

Not sure whether we want to mirror that - would need to keep in sync with their private type declarations.

Reply:
> Not sure whether we want to mirror that - would need to keep in sync with their private type declarations.

And the other issue is that they define some of these private types inside of the `.pyi` files so you can't import the private types.  I think this is too much of a moving target for us to get working now until numpy's typing efforts settle down.

I also believe that numpy is defining some types for typing in Cython code, which makes things even more complex to get working.

For now, I think we should remain incremental on this and accept what I have proposed and then we can investigate how we can better type the interface between pandas and numpy.